### PR TITLE
[flang] optimize array function calls using hlfir.eval_in_mem

### PR DESCRIPTION
**Pull request by jeanPerier**
This patch encapsulate array function call lowering into hlfir.eval_in_mem and allows directly evaluating the call into the LHS when possible.

The conditions are: LHS is contiguous, not accessed inside the function, it is not a whole allocatable, and the function results needs not to be finalized. All these conditions are tested in the previous [hlfir.eval_in_mem optimization patch](https://github.com/llvm/llvm-project/pull/118069) that is leveraging [the extension of getModRef to handle function calls](https://github.com/llvm/llvm-project/pull/117164)).

This yields a 25% speed-up on polyhedron channel2 benchmark (from 1min to 45s measured on an X86-64 Zen 2).

### CODE REVIEW

**Comment by jeanPerier** [comment]
Note that this is a "classic" Fortran compiler optimization, you can verify that at least all of gfortran, ifx, and nvfortran are doing it on something very easy like:

```
subroutine test
interface
 function foo()
    real :: foo(100)
 end function
end interface
   real x(100)
   x = foo()
   call bar(x)
end subroutine
```

File: flang/include/flang/Lower/ConvertCall.h [hunk 1]
```diff
@@ -24,6 +24,10 @@
 
 namespace Fortran::lower {
 
+struct LoweredResult {
+  std::variant<fir::ExtendedValue, hlfir::EntityWithAttributes> result;
+};
```
Comment (tblah) [inline flang/include/flang/Lower/ConvertCall.h:29]: nit: Could this be simplified to
```c++
using LoweredResult = std::variant<fir::ExtendedValue, hlfir::EntityWithAttributes> result;
```

This way callers wouldn't have to add a `.result`.

File: flang/test/HLFIR/order_assignments/where-scheduling.f90 [hunk 2]
```diff
@@ -134,7 +134,7 @@ end function f
 !CHECK-NEXT: run 1 save    : where/mask
 !CHECK-NEXT: run 2 evaluate: where/region_assign1
 !CHECK-LABEL: ------------ scheduling where in _QPonly_once ------------
-!CHECK-NEXT: unknown effect: %{{[0-9]+}} = llvm.intr.stacksave : !llvm.ptr
+!CHECK-NEXT: unknown effect: %11 = fir.call @_QPcall_me_only_once() fastmath<contract> : () -> !fir.array<10x!fir.logical<4>>
```
Comment (tblah) [inline flang/test/HLFIR/order_assignments/where-scheduling.f90:137]: Why did this change?

Comment (tblah) [reply flang/test/HLFIR/order_assignments/where-scheduling.f90:137]: Thanks for the explanation

**Review by tblah — APPROVED** [review]
LGTM with the one nit

File: flang/include/flang/Lower/ConvertCall.h (hunk 1 above)
Comment (jeanPerier) [reply flang/include/flang/Lower/ConvertCall.h]: Updated.

**Comment by jeanPerier** [comment]
Thanks for all the reviews Tom!

**Pull request merged**